datumaro/components/project.py

```diff
@@ -2,486 +2,2481 @@
+
+        for obj in tree.traverse(visit_once=True):
+            if include_files and obj.path not in include_files:
+                continue
+
+            path = osp.join(base_path, obj.path)
+            os.makedirs(osp.dirname(path), exist_ok=True)
+            if obj.type == 'blob':
+                with open(path, 'wb') as f:
+                    obj.stream_data(f)
+            elif obj.type == 'tree':
+                pass
             else:
-                if recursive:
-                    # children items should already be updated
-                    # so we just save them recursively
-                    for source in self._sources.values():
-                        if isinstance(source, ProjectDataset):
-                            source.save(**converter_kwargs)
-
-                self.env.converters.get(DEFAULT_FORMAT).convert(
-                    self.iterate_own(), dataset_save_dir, **converter_kwargs)
-
-            project.save(save_dir)
-        except BaseException:
-            if not save_dir_existed and osp.isdir(save_dir):
-                shutil.rmtree(save_dir, ignore_errors=True)
-            raise
+                raise ValueError("Unexpected object type in a "
+                    "git tree: %s (%s)" % (obj.type, obj.hexsha))
 
     @property
-    def config(self):
-        return self._project.config
+    def head(self) -> str:
+        return self.repo.head.commit.hexsha
 
     @property
-    def env(self):
-        return self._project.env
+    def branch(self) -> str:
+        if self.repo.head.is_detached:
+            return None
+        return self.repo.active_branch
+
+    def rev_parse(self, ref: str) -> Tuple[str, str]:
+        """
+        Expands named refs and tags.
+
+        Returns: object type, object hash
+        """
+        obj = self.repo.rev_parse(ref)
+        return obj.type, obj.hexsha
+
+    def ignore(self, paths: Union[str, List[str]],
+            mode: Union[None, str, IgnoreMode] = None,
+            gitignore: Optional[str] = None):
+        if not gitignore:
+            gitignore = '.gitignore'
+        repo_root = self._project_dir
+        gitignore = osp.abspath(osp.join(repo_root, gitignore))
+        assert is_subpath(gitignore, base=repo_root), gitignore
+
+        _update_ignore_file(paths, repo_root=repo_root,
+            mode=mode, filepath=gitignore)
+
+    HASH_LEN = 40
+
+    @classmethod
+    def is_hash(cls, s: str) -> bool:
+        return len(s) == cls.HASH_LEN
+
+    def log(self, depth=10) -> List[Tuple[Any, int]]:
+        """
+        Returns: a list of (commit, index) pairs
+        """
+
+        commits = []
+
+        if not self.has_commits():
+            return commits
+
+        for commit in zip(self.repo.iter_commits(rev='HEAD'), range(depth)):
+            commits.append(commit)
+        return commits
+
+class DvcWrapper:
+    @staticmethod
+    def import_module():
+        import dvc
+        import dvc.main
+        import dvc.repo
```

Comment:
> I can find no evidence to suggest that this is a supported way to use DVC.

It surely isn't. But the problem with subprocess is performance. It definitely has advantages because it is a separate process etc., but we call DVC commands quite often and even with the current way it is terribly slow. Tests now pass 4-5 times slower.

Reply:
How slowly would the tests run if you used subprocess?

Reply:
`time pytest -v tests/test_project.py`

`subprocess.call()`
41 passed, 3 warnings in 121.98s (0:02:01) 
real    2m2,622s
user    2m40,306s
sys     1m40,964s


`main()`
41 passed, 3 warnings in 51.11s

real    0m51,941s
user    0m16,030s
sys     0m9,094s


The difference is more than 2x.